File: python/apps/SimpleSiggen.py
#!/usr/bin/env python
"""Simple signal generator for testing transmit

Continuously output a carrier with single sideband sinusoid amplitude
modulation.

Terminate with cntl-C.
"""

import argparse
import math
import signal
import time

import numpy as np

import SoapySDR
from SoapySDR import * #SOAPY_SDR_ constants

import soapy_log_handle

def siggen_app(
        args,
        rate,
        ampl=0.7,
        freq=None,
        tx_bw=None,
        tx_chan=0,
        tx_gain=None,
        tx_ant=None,
        clock_rate=None,
        wave_freq=None
):
    """Generate signal until an interrupt signal is received."""

    if wave_freq is None:
        wave_freq = rate / 10

    sdr = SoapySDR.Device(args)
    #set clock rate first
    if clock_rate is not None:
        sdr.setMasterClockRate(clock_rate)

    #set sample rate
    sdr.setSampleRate(SOAPY_SDR_TX, tx_chan, rate)
    print("Actual Tx Rate %f Msps"%(sdr.getSampleRate(SOAPY_SDR_TX, tx_chan) / 1e6))

    #set bandwidth
    if tx_bw is not None:
        sdr.setBandwidth(SOAPY_SDR_TX, tx_chan, tx_bw)

    #set antenna
    print("Set the antenna")
    if tx_ant is not None:
        sdr.setAntenna(SOAPY_SDR_TX, tx_chan, tx_ant)

    #set overall gain
    print("Set the gain")
    if tx_gain is not None:
        sdr.setGain(SOAPY_SDR_TX, tx_chan, tx_gain)

    #tune frontends
    print("Tune the frontend")
    if freq is not None:
        sdr.setFrequency(SOAPY_SDR_TX, tx_chan, freq)


    print("Create Tx stream")
    tx_stream = sdr.setupStream(SOAPY_SDR_TX, SOAPY_SDR_CF32, [tx_chan])
    print("Activate Tx Stream")
    sdr.activateStream(tx_stream)
    phase_acc = 0
    phase_inc = 2*math.pi*wave_freq/rate
    stream_mtu = sdr.getStreamMTU(tx_stream)
    samps_chan = np.array([ampl]*stream_mtu, np.complex64)

    time_last_print = time.time()
    total_samps = 0

    state = dict(running=True)

    def signal_handler(signum, _):
        print('Signal handler called with signal {}'.format(signum))
        state['running'] = False

    signal.signal(signal.SIGINT, signal_handler)

    while state['running']:
        phase_acc_next = phase_acc + stream_mtu*phase_inc
        phases = np.linspace(phase_acc, phase_acc_next, stream_mtu)
        samps_chan = ampl*np.exp(1j * phases).astype(np.complex64)
        phase_acc = phase_acc_next
        while phase_acc > math.pi * 2:
            phase_acc -= math.pi * 2

        status = sdr.writeStream(tx_stream, [samps_chan], samps_chan.size, timeoutUs=1000000)
        if status.ret != samps_chan.size:
            raise Exception("Expected writeStream() to consume all samples! %d" % status.ret)
        total_samps += status.ret

        if time.time() > time_last_print + 5.0:
            rate = total_samps / (time.time() - time_last_print) / 1e6
            print("Python siggen rate: %f Msps" % rate)
            total_samps = 0
            time_last_print = time.time()

    #cleanup streams
    print("Cleanup stream")
    sdr.deactivateStream(tx_stream)
    sdr.closeStream(tx_stream)
    print("Done!")

def main():
    """Parse command line arguments and start sig-gen."""
    parser = argparse.ArgumentParser(
        description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter)

    parser.add_argument("--args", type=str, help="device factor arguments", default="")
    parser.add_argument("--rate", type=float, help="Tx and Rx sample rate", default=1e6)
    parser.add_argument("--ampl", type=float, help="Tx digital amplitude rate", default=0.7)
    parser.add_argument("--tx-ant", type=str, help="Optional Tx antenna")
    parser.add_argument("--tx-gain", type=float, help="Optional Tx gain (dB)")
    parser.add_argument("--tx-chan", type=int, help="Transmitter channel (def=0)", default=0)
    parser.add_argument("--freq", type=float, help="Optional Tx and Rx freq (Hz)")
    parser.add_argument("--tx-bw", type=float, help="Optional Tx filter bw (Hz)", default=5e6)
    parser.add_argument("--wave-freq", type=float, help="Baseband waveform freq (Hz)")
    parser.add_argument("--clock-rate", type=float, help="Optional clock rate (Hz)")
    parser.add_argument("--debug", action='store_true', help="Output debug messages")
    parser.add_argument(
        "--abort-on-error", action='store_true',
        help="Halts operations if the SDR logs an error")

    options = parser.parse_args()

    if options.abort_on_error:
        exception_level = SOAPY_SDR_WARNING
    else:
        exception_level = None
    soapy_log_handle.set_python_log_handler(exception_level=exception_level)
    if options.debug:
        SoapySDR.setLogLevel(SOAPY_SDR_DEBUG)

    siggen_app(
        args=options.args,
        rate=options.rate,
        ampl=options.ampl,
        freq=options.freq,
        tx_bw=options.tx_bw,
        tx_ant=options.tx_ant,
        tx_gain=options.tx_gain,
        tx_chan=options.tx_chan,
        clock_rate=options.clock_rate,
        wave_freq=options.wave_freq,
    )

if __name__ == '__main__':
    main()

File: luajit/apps/SimpleSiggen.lua
#!/usr/bin/env luajit
-- Copyright (c) 2021-2022 Nicholas Corgan
-- SPDX-License-Identifier: BSL-1.0

local SoapySDR = require("SoapySDR")

local ffi = require("ffi")

local argparse = require("argparse")
local complex = require("complex")
local signal = require("posix.signal")

local math = require("math")
local os = require("os")

--
--
--

local function toboolean(str)
    return (string.lower(str) == "true")
end

local parser = argparse("SimpleSiggen")
parser:option("--args", "Device args", "")
parser:option("--rate", "Sample rate", 1e6):convert(tonumber)
parser:option("--ampl", "TX digital amplitude", 0.7):convert(tonumber)
parser:option("--tx-ant", "Optional TX antenna")
parser:option("--tx-gain", "Optional TX gain"):convert(tonumber)
parser:option("--tx-chan", "TX channel", 0):convert(tonumber)
parser:option("--freq", "Optional frequency (Hz)"):convert(tonumber)
parser:option("--tx-bw", "Optional TX filter bandwidth (Hz)", 5e6):convert(tonumber)
parser:option("--wave-freq", "Optional wave frequency (Hz)"):convert(tonumber)
parser:option("--clock-rate", "Optional clock rate (Hz)"):convert(tonumber)
parser:option("--debug", "Output debug messages", false):convert(toboolean)

local args = parser:parse()

local function cleanup(sdr, stream)
    local function __cleanup(sdr, stream)
        if sdr ~= nil and stream ~= nil then
            if args.debug then print("Deactivating and closing stream"); end
            sdr:deactivateStream(stream)
            sdr:closeStream(stream)
        end
    end

    print("Cleanup...")
    pcall(__cleanup, sdr, stream)
end

local function linspace(startval, endval, steps)
    interval = (endval / math.abs(endval)) * math.abs(endval - startval) / (steps - 1)
    local ret = {}

    for i = 1,steps do
        ret[i] = startval + (steps[i] * interval)
    end

    return ret
end

local function getRealAsComplexArray(real, len)
    local ret = {}
    for i = 1,len do
        ret[i] = ffi.new("complex", real, 0)
    end
    return ret
end

local function populateSampsWithReal(real, len, samps)
    for i = 0,len-1 do
        samps[i] = ffi.new("complex", real, 0)
    end
end

local function getSamp(ampl, phase)
    local lcomplexRet = complex.new(ampl, 0) * (complex.I * phase):exp()
    return ffi.new("complex", lcomplexRet:real(), lcomplexRet:imag())
end

local function populateSamps(ampl, phases, samps)
    local ret = {}
    for i = 0,#phases-1 do
        samps[i] = getSamp(ampl, phases[i+1])
    end
end

--
--
--

-- Configure logger
local function logFunction(logLevel, message)
    print(string.format("%s: %s", os.date("%Y-%m-%d %H:%M:%S"), message))
end

local logLevel
if args.debug then logLevel = SoapySDR.Logger.Level.DEBUG
else               logLevel = SoapySDR.Logger.Level.WARNING
end
SoapySDR.Logger.registerHandler(logFunction)
SoapySDR.Logger.setLevel(logLevel)

-- Initialize device
local sdr = SoapySDR.Device.new(args.args)
print("----------")
print(string.format("Device:  %s", sdr))
print(string.format("Channel: %d", args.tx_chan))

-- Set master clock rate
if args.clock_rate then
    print("----------")
    print(string.format("Setting master clock rate: %f Hz", args.clock_rate))
    sdr:setMasterClockRate(args.clock_rate)
    print(string.format("Actual:                    %f Hz", sdr:getMasterClockRate()))
end

-- Set sample rate
print("----------")
print(string.format("Setting TX sample rate: %f Hz", args.rate))
sdr:setSampleRate(SoapySDR.Direction.TX, args.tx_chan, args.rate)
print(string.format("Actual:                 %f Hz", sdr:getSampleRate(SoapySDR.Direction.TX, args.tx_chan)))

-- Set bandwidth
print("----------")
print(string.format("Setting TX bandwidth: %f Hz", args.tx_bw))
sdr:setBandwidth(SoapySDR.Direction.TX, args.tx_chan, args.tx_bw)
print(string.format("Actual:               %f Hz", sdr:getBandwidth(SoapySDR.Direction.TX, args.tx_chan)))

-- Set antenna
if args.tx_ant then
    print("----------")
    print(string.format("Setting TX antenna: %s", args.tx_ant))
    sdr:setAntenna(SoapySDR.Direction.TX, args.tx_chan, args.tx_ant)
    print(string.format("Actual:             %s", sdr:getAntenna(SoapySDR.Direction.TX, args.tx_chan)))
end

-- Set overall gain
if args.tx_gain then
    print("----------")
    print(string.format("Setting TX gain: %f dB", args.tx_gain))
    sdr:setGain(SoapySDR.Direction.TX, args.tx_chan, args.tx_gain)
    print(string.format("Actual:          %f dB", sdr:getGain(SoapySDR.Direction.TX, args.tx_chan)))
end

-- Tune frontend
if args.freq then
    print("----------")
    print(string.format("Setting TX frequency: %f Hz", args.freq))
    sdr:setFrequency(SoapySDR.Direction.TX, args.tx_chan, args.freq)
    print(string.format("Actual:               %f Hz", sdr:getFrequency(SoapySDR.Direction.TX, args.tx_chan)))
end

-- Initialize stream
local stream = sdr:setupStream(SoapySDR.Direction.TX, SoapySDR.Format.CF32, {args.tx_chan})
if stream == nil then
    error("Failed to set up TX stream")
end

-- Activate stream
local errorCode = sdr:activateStream(stream)
if errorCode ~= 0 then
    error("activateStream returned " .. SoapySDR.ErrorToString(errorCode))
end
local streamMTU = sdr:getStreamMTU(stream)

-- Set up samples
local samps = ffi.new("complex float[?]", mtu)
local samps2D = ffi.new("complex float*[1]", {samps})
local timeoutUs = 1000000

local waveFreq = args.wave_freq or (args.rate / 10)
local phaseAcc = 0
local phaseInc = 2 * math.pi * waveFreq / args.rate
populateSampsWithReal(ampl*streamMTU, streamMTU, samps)

-- Set up signal handling
local running = true

local function sighandler(signo)
    running = false
end

signal.signal(signal.SIGINT, sighandler)

-- Run siggen until interrupted
local totalSamps = 0
local lastTime = os.clock()

while running do
    local phaseAccNext = phaseAcc + (streamMTU * phaseInc)
    local phases = linspace(phaseAcc, phaseAccNext, streamMTU)
    populateSamps(ampl, phases, samps)
    phaseAcc = phaseAccNext

    while phaseAcc > (math.pi * 2) do
        phaseAcc = (phaseAcc - (math.pi * 2))
    end

    local writeStreamRet = sdr:writeStream(stream, samps2D, streamMTU, timeoutUs)
    if writeStreamRet < 0 then
        cleanup(sdr, stream)
        error(string.format("writeStream() returned error: %s", SoapySDR.Error.ToString(writeStreamRet)))
    elseif writeStreamRet ~= streamMTU then
        cleanup(sdr, stream)
        error(string.format("Expected writeStream() to consume all samples. Consumed %d.", writeStreamRet))
    end

    totalSamps = totalSamps + writeStreamRet
    local now = os.clock()
    local duration = now - lastTime
    if duration >= 5 then
        local rate = totalSamps / duration / 1e6
        print(string.format("LuaJIT siggen rate: %f Msps", rate))
        totalSamps = 0
        lasTime = now
    end
end

-- Clean up
cleanup(fd, sdr, stream)
print("Done!")
